### PR TITLE
🙈 Used `got` to handle requests for image-size

### PR DESCRIPTION
**Pull request by aileen**
refs #8589
refs #8868

Swap `request` with `got` in `getImageSizeFromUrl` fn.

TODOs:
- [x] make redirects work
- [x] fix tests
- [x] check possible fix for #8885 -> done with #8915

### CODE REVIEW

File: core/test/unit/server_helpers/ghost_head_spec.js [hunk 1]
```diff
@@ -42,7 +42,7 @@ describe('{{ghost_head}} helper', function () {
                 return localSettingsCache[key];
             });
 
-            configUtils.set('url', 'http://localhost:82832/');
+            configUtils.set('url', 'http://localhost:65530/');
```
Comment (AileenCGN) [inline core/test/unit/server_helpers/ghost_head_spec.js]: I had to change the port for testing here, as `got` only works with ports < `65539` and we're not stubbing the `getImageSize` fn yet.

**Comment by AileenCGN** [comment]
@ErisDS: I assume this change needs a lot of testing, as our image size util is the cause for many dispatcher issues. If you feel like we're missing some more tests, please let me know.

**Comment by AileenCGN** [comment]
Also, I didn't test yet, if this PR solves the issues in #8885 and #8589.

**Comment by yanke-guo** [comment]
@AileenCGN @ErisDS I made a small work around PR #8915 for issue #8885, just handle the situation when 'icon' is a full url.

File: core/server/utils/image-size-from-url.js [hunk 2]
```diff
@@ -34,7 +31,7 @@ var sizeOf       = require('image-size'),
 module.exports.getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
     return new Promise(function imageSizeRequest(resolve, reject) {
```
Comment (ErisDS) [inline core/server/utils/image-size-from-url.js]: As `got` uses promises, I don't think this needs to be wrapped anymore?

File: core/server/utils/image-size-from-url.js [hunk 3]
```diff
@@ -51,71 +48,56 @@ module.exports.getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
...
+        };
 
-        requestHandler = imagePath.indexOf('https') === 0 ? https : http;
-        options.headers = {'User-Agent': 'Mozilla/5.0'};
+        got(
+            imagePath,
+            requestOptions
+        ).then(function (response) {
+            try {
+                dimensions = sizeOf(response.body);
```
Comment (ErisDS) [inline core/server/utils/image-size-from-url.js]: could use promisify here, to make this use the promise signature  we use elsewhere, that would further reduce the number of error cases

File: core/server/utils/image-size-from-url.js [hunk 4]
```diff
@@ -51,71 +48,56 @@ module.exports.getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
...
+                    statusCode: err.statusCode,
+                    context: err.url
+                }));
+            } else {
+                return reject(new errors.InternalServerError({
+                    message: 'Unknown Request error.',
+                    code: 'IMAGE_SIZE',
+                    statusCode: err.statusCode,
+                    context: err.url
+                }));
```
Comment (ErisDS) [inline core/server/utils/image-size-from-url.js]: Yay for less error cases 🎉

File: core/server/utils/image-size-from-url.js [hunk 5]
```diff
@@ -51,71 +48,56 @@ module.exports.getImageSizeFromUrl = function getImageSizeFromUrl(imagePath) {
...
+                    context: imagePath.href
+                }));
+            }
+        }).catch(function (err) {
+            if (err.statusCode === 404) {
+                return reject(new errors.NotFoundError({
+                    message: 'Image not found.',
+                    code: 'IMAGE_SIZE',
+                    statusCode: err.statusCode,
+                    context: err.url
```
Comment (ErisDS) [inline core/server/utils/image-size-from-url.js]: does err always have a url? Could we not use imagePath.href here?

**Review by ErisDS — CHANGES_REQUESTED** [review]
This is looking so much better! I think there is a last pass through to do to tidy up one or two more things :)

File: core/server/utils/image-size-from-url.js (hunk 5 above)
Comment (AileenCGN) [inline core/server/utils/image-size-from-url.js]: Hmm... in my tests, `err.url` has always been populated by `got`, but seeing [here](https://github.com/sindresorhus/got#errors) that is says "if available", I probably should return a conditional just to make sure.

File: core/server/utils/image-size-from-url.js (hunk 3 above)
Comment (AileenCGN) [inline core/server/utils/image-size-from-url.js]: I tried refactoring as you recommended, but I couldn't make it work... Guess I need some more time or assistance with this 😔

Comment (AileenCGN) [reply core/server/utils/image-size-from-url.js]: After trying some more time to promisify the `sizeOf` fn, I figured that it won't work asynchronously when working with the buffer. See the note [here](https://github.com/image-size/image-size#asynchronous).

So IMO we have three options:
1. Leave the whole `fn` wrapped in a promise like it is now
2. Use the URL and not the buffer (which will then again not work when reading from the local file storage as planned in #8900)
3. Make it work synchronously and just wrap it in a `try`

Comment (kirrg001) [reply core/server/utils/image-size-from-url.js]: I would leave this as it is. Your comment in the code base makes it very clear 👍

File: core/server/utils/image-size-from-url.js (hunk 2 above)
Comment (kirrg001) [reply core/server/utils/image-size-from-url.js]: Yeah this should work 👍

**Comment by kirrg001** [comment]
I've tested several cases from the past and looks like `got` is doing a good job.
e.g. follow redirects, timeouts

File: core/server/utils/image-size-from-url.js (hunk 2 above)
Comment (AileenCGN) [reply core/server/utils/image-size-from-url.js]: Alright! Finally got it working! Change is pushed ;)

**Review by kirrg001 — APPROVED** [review]

**Pull request merged**